src/megaapi_impl.cpp

```diff
@@ -25146,12 +25057,176 @@ void MegaFolderUploadController::onTransferFinish(MegaApi *, MegaTransfer *t, Me
+        {
+            // lambda function that will be executed as completion function in putnodes procresult
+            mLastError = e;
+            mPendingFolders -= nn.size();
+
+            if (!e)
+            {
+                startFileUploads();
+            }
+            else
+            {
+                mIncompleteTransfers++;
+            }
+            checkCompletion();
+        });
+    }
+}
+
+void MegaFolderUploadController::startFileUploads()
+{
+    auto itFiles = mPendingFiles.begin();
+    while (itFiles != mPendingFiles.end())
+    {
+       handle parentHandle = itFiles->second;
+       auto it = client->mTempHandleToNodeHandle.find(parentHandle);
+       unique_ptr <MegaNode> parentNode((it != client->mTempHandleToNodeHandle.end())
+               ? megaApi->getNodeByHandle(it->second)     // new node has been created, so we need to retrieve it's handle
+               : megaApi->getNodeByHandle(parentHandle)); // node already existed, so get it's handle from mPendingFiles
+
+       if (parentNode != nullptr)
+       {
+            pendingTransfers++;
+            const LocalPath &localpath = itFiles->first;
+            FileSystemType fsType = client->fsaccess->getlocalfstype(localpath);
+            megaApi->startUpload(false, localpath.toPath(*client->fsaccess).c_str(), parentNode.get(), (const char *)NULL, -1, tag, false, NULL, false, false, fsType, this);
+            itFiles = mPendingFiles.erase(itFiles);
+       }
+       else
+       {
+           ++itFiles;
+       }
+    }
+}
+
+void MegaFolderUploadController::checkCompletion()
+{
+    if (!cancelled && !recursive && !pendingTransfers && !mPendingFolders)
+    {
+        LOG_debug << "Folder transfer finished - " << transfer->getTransferredBytes() << " of " << transfer->getTotalBytes();
+        mFolderStructure.clear();
+        mPendingFiles.clear();
+        transfer->setState(MegaTransfer::STATE_COMPLETED);
+        transfer->setLastError(&mLastError);
+        DBTableTransactionCommitter committer(client->tctable);
+        megaApi->fireOnTransferFinish(transfer, make_unique<MegaErrorPrivate>(!mIncompleteTransfers ? API_OK : API_EINCOMPLETE), committer);
+    }
+}
+
+handle MegaFolderUploadController::addNewNodeToVector(handle targetHandle, handle parentHandle, const char * folderName)
+{
+    handle newNodeHandle;
+    NewNode newnode;
+    SymmCipher key;
+    string attrstring;
+    byte buf[FOLDERNODEKEYLENGTH];
+
+    // set up new node as folder node
+    newnode.source = NEW_NODE;
+    newnode.type = FOLDERNODE;
+    newnode.nodehandle = 0;
+    newnode.parenthandle = UNDEF;
+
+    // generate fresh random key for this folder node
+    client->rng.genblock(buf,FOLDERNODEKEYLENGTH);
+    newnode.nodekey.assign((char*)buf,FOLDERNODEKEYLENGTH);
+    key.setkey(buf);
+
+    // generate fresh attribute object with the folder name
+    AttrMap attrs;
+    string sname = folderName;
+    client->fsaccess->normalize(&sname);
```

Comment:
You can use MegaClient::putnodes_prepareOneFolder() to do a lot of this setup, and save repeating code here.  thanks